### PR TITLE
feat: make wallet password optional (agent-friendly)

### PR DESCRIPTION
**Pull request by askeluv**
## Problem

AI agents create wallets interactively, pick a password, then lose it on session restart. This is a fundamental issue for agent-first CLIs — agents are amnesiac by nature.

## Solution

Make wallet passwords **optional by default**. Keys are stored unencrypted (like `~/.ssh/` keys). Users who want encryption at rest opt in via `NANSEN_WALLET_PASSWORD` env var.

### How it works

- `nansen wallet create my-wallet` → no password, keys stored as plaintext JSON
- `NANSEN_WALLET_PASSWORD=xyz nansen wallet create my-wallet` → encrypted with that password
- All other commands (export, delete, execute, send) auto-detect whether the wallet is password-protected
- If existing wallets have a password, new wallets require the env var too (consistency)
- No interactive password prompts anywhere in wallet create

### Security model

Same as SSH keys — the filesystem is the security boundary. Agents running in sandboxed containers (Docker, cloud VMs) get isolation from the runtime. Humans who want defense-in-depth set the env var in their shell profile or secrets manager.

## Changes

- `src/wallet.js`: Optional password in create, export, delete; plaintext encrypt/decrypt path
- `src/trading.js`: Auto-detect passwordless wallets in execute flow
- `src/transfer.js`: Skip password verification for unprotected wallets
- `src/x402.js`: Handle passwordless wallet flow

## Tests

All 135 existing tests pass (wallet: 25, trading: 55, transfer: 55).

### CODE REVIEW

File: src/wallet.js [hunk 1]
```diff
@@ -78,6 +78,11 @@ function deriveKey(password, salt) {
  * Returns a JSON-serializable object with all params needed for decryption.
  */
 export function encryptKey(privateKeyHex, password) {
+  // No password → store key as plaintext wrapper
+  if (!password) {
```
Comment (0xlaveen) [inline src/wallet.js:82]: Nit: the field name `plaintext` is a bit on-the-nose — if someone `cat`s the wallet file, it literally advertises "here is the plaintext key." SSH doesn't label key files as "PLAINTEXT PRIVATE KEY" either.

Consider a less descriptive name like `key` or `data`, with a sibling marker:
```js
if (!password) {
  return { data: privateKeyHex, encrypted: false };
}
```

Not a blocker, just security hygiene.

File: src/x402.js [hunk 2]
```diff
@@ -96,18 +96,21 @@ export async function* createPaymentSignatures(response, url, options = {}) {
...
-
-  let exportWallet, listWallets;
+  let password = options.password || process.env.NANSEN_WALLET_PASSWORD || null;
+  // For passwordless wallets, we still need to proceed
+  let walletMod;
   try {
-    const walletMod = await import('./wallet.js');
-    exportWallet = walletMod.exportWallet;
-    listWallets = walletMod.listWallets;
+    walletMod = await import('./wallet.js');
```
Comment (0xlaveen) [inline src/x402.js]: The restructure here is slightly fragile — `exportWallet` and `listWallets` are destructured after the try/catch, but if the import fails it already returns early. Can simplify to a single block:

```js
try {
  const { exportWallet, listWallets, getWalletConfig } = await import('./wallet.js');
  const config = getWalletConfig();
  if (config.passwordHash && !password) return;
  // ... use exportWallet, listWallets directly below
} catch { return; }
```

Avoids the dangling `let walletMod` and the separate re-assignment.

**Comment by Codier** [comment]
@claude Please review this PR and provide your assessment.

File: src/wallet.js [hunk 3]
```diff
@@ -78,6 +78,11 @@ function deriveKey(password, salt) {
  * Returns a JSON-serializable object with all params needed for decryption.
  */
 export function encryptKey(privateKeyHex, password) {
+  // No password → store key as unencrypted wrapper
+  if (!password) {
```
Comment (Codier) [inline src/wallet.js:82]: 🔴 **Blocking — Security**: `decryptKey` trusts `encryptedData.encrypted === false` without any other validation. If an attacker can modify the wallet JSON file on disk (or if a bug writes a malformed file), they could set `encrypted: false` on an *actually encrypted* wallet and `data` to an arbitrary key, bypassing encryption entirely.

Consider adding a secondary check — e.g. store a HMAC of the plaintext key alongside it so you can at least verify integrity, or validate that `encryptedData` has the expected shape for each path (plaintext must not have `salt`/`iv`/`authTag`, and vice versa).

File: src/wallet.js [hunk 4]
```diff
@@ -400,7 +416,7 @@ export function exportWallet(name, password) {
   }
 
   const config = getWalletConfig();
-  if (!verifyPassword(password, config)) {
```
Comment (Codier) [inline src/wallet.js:403]: 🟡 **Should Fix**: The `⚠️ No NANSEN_WALLET_PASSWORD set` warning goes to `log()` which is stdout. For structured JSON output (hard rule), this warning string will break JSON parsers. The wallet creation result JSON should include a `"warning"` field instead, and the emoji warning should go to stderr.

File: src/wallet.js [hunk 5]
```diff
@@ -103,6 +108,11 @@ export function encryptKey(privateKeyHex, password) {
  * @throws {Error} If password is wrong
  */
```
Comment (Codier) [inline src/wallet.js:109]: 🟡 **Should Fix**: The CLI handler checks `config.passwordHash && !password` and exits, but the `createWallet()` function itself does NOT enforce this. Any direct caller of `createWallet(name, null)` will happily create an unencrypted wallet even when other wallets are encrypted. The guard should live in the function, not just the CLI handler.

File: src/__tests__/wallet.test.js [hunk 6]
```diff
@@ -92,6 +92,16 @@ describe('encryption', () => {
 
     expect(() => decryptKey(encrypted, 'wrong-password')).toThrow('Incorrect password');
   });
+
+  it('encryptKey with null password returns unencrypted wrapper', () => {
+    const result = encryptKey('deadbeef', null);
+    expect(result).toEqual({ data: 'deadbeef', encrypted: false });
+  });
+
+  it('decryptKey handles unencrypted wrapper', () => {
```
Comment (Codier) [inline src/__tests__/wallet.test.js:100]: 🟢 **Nitpick**: Good that you added tests for the null-password path. Consider also testing the negative case: `decryptKey` called with an encrypted blob but null password should still throw, and calling with `{encrypted: false, data: ...}` that also has `salt`/`iv` fields should probably be rejected.

File: src/x402.js [hunk 7]
```diff
@@ -96,14 +96,14 @@ export async function* createPaymentSignatures(response, url, options = {}) {
   const ranked = rankRequirements(requirements);
   if (ranked.length === 0) return;
 
-  const password = options.password || process.env.NANSEN_WALLET_PASSWORD;
-  if (!password) return;
-
+  const password = options.password || process.env.NANSEN_WALLET_PASSWORD || null;
   let exportWallet, listWallets;
```
Comment (Codier) [inline src/x402.js]: 🟡 **Should Fix**: When `config.passwordHash && !password`, you silently `return` — the payment just fails without any signal to the caller. At minimum yield a diagnostic or log to stderr so the user knows *why* x402 payment was skipped.

**Comment by Codier** [comment]
Overall having no password for the wallet feels like a security risk here. I would really hope to get @mrh-nansen to at least approve the overall concept before moving further on this PR
